### PR TITLE
Filter completion results - #41

### PR DESCRIPTION
**Pull request by lhein**

### CODE REVIEW

**Comment by lhein** [comment]
fixes #41

File: src/main/java/com/github/cameltooling/lsp/internal/completion/CamelOptionNamesCompletionsFuture.java [hunk 1]
```diff
@@ -27,20 +27,23 @@
...
 		this.camelComponentName = camelComponentName;
 		this.isProducer = isProducer;
+		this.filterString = filterText;
 	}
 
 	@Override
 	public List<CompletionItem> apply(CamelCatalog catalog) {
 		Stream<EndpointOptionModel> endpointOptions = ModelHelper.generateComponentModel(catalog.componentJSonSchema(camelComponentName), true).getEndpointOptions().stream();
 		return endpointOptions
+				.filter(parameter -> filterString != null && filterString.trim().length()>0 ? parameter.getName().startsWith(filterString) : true)
```
[the remark this answers is not in the document]
Comment (lhein) [reply src/main/java/com/github/cameltooling/lsp/internal/completion/CamelOptionNamesCompletionsFuture.java]: right. I added it in the componentschemefuture too now.

**Comment by lhein** [comment]
> the filter applied is using the whole string and doesn't take care of cursor position, so when placing the cursor at second position for instance although there are more characters typed, it provides results for the whole characters and not from where the cursor is:

Yes, that is why I asked you before what the reason was to merge completions for path and scheme into the same future. I would have expected those would be split into separate futures.

**Comment by lhein** [comment]
> As Eclipse was not providing very heuristic, it might be better but on VSCode, we will loose some features like:

No idea what you are talking about and the picture is not really helping here.

**Comment by lhein** [comment]
> found another usecase even more problematic; it is no more proposing all the available values:

Agreed. That is a bug.

File: src/main/java/com/github/cameltooling/lsp/internal/completion/CamelOptionNamesCompletionsFuture.java (hunk 1 above)
Comment (apupier) [inline src/main/java/com/github/cameltooling/lsp/internal/completion/CamelOptionNamesCompletionsFuture.java]: this is duplicated code across several classes, might worth extracting a Predicate from it

**Comment by apupier** [comment]
> As Eclipse was not providing very heuristic, it might be better but on VSCode, we will loose some features like:
> 
> No idea what you are talking about and the picture is not really helping here.

VSCode is doing a sort of the proposals and highlighting some of them. We will have a single result with the proposal.

**Comment by apupier** [comment]
> the filter applied is using the whole string and doesn't take care of cursor position, so when placing the cursor at second position for instance although there are more characters typed, it provides results for the whole characters and not from where the cursor is:
> 
> Yes, that is why I asked you before what the reason was to merge completions for path and scheme into the same future. I would have expected those would be split into separate futures.

I don't think that it is related. The whole existing string is used for all part of the Camel URI, instead of just the part before the cursor. The same apply for all parts of the Camel URI.

**Comment by lhein** [comment]
> VSCode is doing a sort of the proposals and highlighting some of them. We will have a single result with the proposal.

Isn't that the whole point of filtering, Aurelien? Reducing the resultset by excluding values which doesn't match what typed?

On the other hand, if you mark the "timer" backwards so the cursor is at the beginning then you get your full list back ;)

**Comment by lhein** [comment]
I think the tests should be now more like the one u r looking for.

**Comment by apupier** [comment]
> I think the tests should be now more like the one u r looking for.

unless I'm missing something, the use case mentioned in this comment https://github.com/camel-tooling/camel-language-server/pull/54#discussion_r184608017 is still not available

**Pull request merged**